### DISCUSSION TITLE
Using ContactListener for Virtual Character

### DISCUSSION THREAD

**Discussion by ZyChua2000** [jrouwe/JoltPhysics#1363] [Q&A]
I have a Virtual Character in my scene.

ContactListener currently works between other non-character rigid bodies, inclusive of sensors, dynamic bodies, kinematic bodies and static bodies.

I've previously tried to use CharacterContactListener for my character, but I realised that CharacterContactListener only has an OnContactAdded.

However, for the purpose of my game engine, I will need OnContactAdded, OnContactPersisted and OnContactRemoved as per the ContactListener class. Hence, I attempted to use an inner rigid body in my virtual character, which was successfully created as I have seen from the debug renderer for bodies. However, the inner rigidbody doesn't seem to trigger the ContactListener. May I check if there's any ways to allow my character to receive persisted and removed callbacks for collision?

**Comment by jrouwe** [comment]
> However, for the purpose of my game engine, I will need OnContactAdded, OnContactPersisted and OnContactRemoved as per the ContactListener class.

Please take a look at this thread here: https://github.com/jrouwe/JoltPhysics/issues/1027

> However, the inner rigidbody doesn't seem to trigger the ContactListener. May I check if there's any ways to allow my character to receive persisted and removed callbacks for collision?

This works fine for me. I adjusted the CharacterVirtualTest sample to show that it works (you'll see a green box appear around the character on entering the sensor): [Character.zip](https://github.com/user-attachments/files/17957810/Character.zip)

**Comment by ZyChua2000** [comment]
Ah yes, it works now, apparently I mixed up some of the layers. Thank you
